blog/_posts/2023-01-16-wsl2spackapptainer.md

```diff
@@ -0,0 +1,284 @@
+
+    version("walberla", branch="walberla")
+
+    depends_on("cmake@3.0:", type="build")
+    depends_on("mpi")
+    depends_on("boost+serialization+filesystem+system+python+mpi")
+
+    # TODO: replace this with an explicit list of components of Boost,
+    # for instance depends_on('boost +filesystem')
+    # See https://github.com/spack/spack/pull/22303 for reference
+    depends_on(Boost.with_default_variants)
+    extends("python")
+    depends_on("py-cython@0.23:", type="build")
+    depends_on("py-numpy", type=("build", "run"))
+    depends_on("fftw")
+    depends_on("hdf5+hl+mpi")
+```
+
+That includes everything but turning on the waLBerla support, so we add in:
+
+```python
+    variant("walberla", default=True, description="Include waLBerla support")
+    def cmake_args(self):
+        spec = self.spec
+        args = [
+            self.define_from_variant("ESPRESSO_BUILD_WITH_WALBERLA", "walberla")
+        ]
+        return args
+```
+
+Have we arrived at working build?  `spack install py-espresso-walberla` rather does make it look like we have.
+
+Now was that really the right way to add this new version of py-espresso?  I'm not going to lose sleep over it right now.
+
+## Putting this in a container
+
+What if I wanted to export this as a container to run on another system?  This is where I think we can plug Apptainer and Spack together.
+
+### A basic Spack container recipe
+
+Spack has support for making Apptainer compatible recipes for exporting Spack environments into containers, but it's not quite up to the challenge of working with this modified Spack (that exists on this machine alone).
+
+So let's do the nearly-but-not quite, then tweak it.  If we write a spack.yaml file:
+```yaml
+spack:
+  specs:
+  - py-espresso-walberla
+
+  container:
+    format: singularity
+    images:
+      os: "ubuntu:22.04"
+      spack: develop
+```
+We can then ask spack to generate a recipe:
+```bash
+spack containerize > py-espresso-walberla.def
+```
+
+That tries to use standard Spack, not our version, so let's just tweak it to delete the version in the container, and copy ours in instead:
+```bash
+%files
+spack /opt/newspack
+
+%post
+   rm -rf /opt/spack ~/.spack
+   mv /opt/newspack /opt/spack
+```
+
+Installing apptainer, and required dependencies:
+```bash
+sudo apt-get install -y uidmap squashfuse fuse2fs fuse-overlayfs gfortran
+spack install apptainer@1.1.5~suid
+```
+
+Then use it to build the container:
+```bash
+spack load apptainer
+apptainer build -B /run py-espresso-walberla.sif py-espresso-walberla.def
+```
+
+There's a couple of gotchas in there that I seem to have magically sidestepped with awesome foresight.  Ahem.  Make sure you don't use the default OS image, else the compiler used is too old, and we've not done anything to load a newer compiler.  Also, if you get a crypic permission denied error, that's what the `-B /run` avoids and you might end up spending far too long scratching you head why it's complaining.
```

Comment:
```suggestion
There's a couple of gotchas in there that I seem to have magically sidestepped with awesome foresight.  Ahem.  Make sure you don't use the default OS image, else the compiler used is too old, and we've not done anything to load a newer compiler.  Also, if you get a cryptic permission denied error, that's what the `-B /run` avoids and you might end up spending far too long scratching you head why it's complaining.
```

Could explain the `-B` bit in a bit more detail if so inclined!

Reply:
Done in 0390226